File: src/services/calendar.py
"""
Calendarific API service
Provides holiday information to help adjust restaurant recommendations
API Documentation: https://calendarific.com/api-documentation
"""

from typing import TypedDict, List, Optional, Union
import requests
from datetime import datetime


class Holiday(TypedDict):
    """Holiday information"""
    name: str
    description: str
    date: dict
    type: str
    locations: List[str]


class HolidaysResponse(TypedDict):
    """Normalized holidays response format"""
    holidays: List[Holiday]
    isHoliday: bool
    holidayNames: List[str]


def get_holidays(
    api_key: str,
    country_code: str,
    year: int,
    month: Optional[int] = None,
    day: Optional[int] = None
) -> HolidaysResponse:
    """
    Gets holidays for a specific country and date/year
    
    Args:
        api_key: Calendarific API key from environment variable
        country_code: ISO 3166-1 alpha-2 country code (e.g., 'US', 'GB', 'CA')
        year: Year to get holidays for
        month: Optional month (1-12)
        day: Optional day (1-31)
    
    Returns:
        Holidays data for the specified date/year
    
    Raises:
        Exception: If API request fails or API key is missing
    """
    if not api_key:
        raise ValueError('Calendarific API key is required. Set CALENDARIFIC_API_KEY environment variable.')
    
    if not country_code or not year:
        raise ValueError('Country code and year are required.')
    
    url = 'https://calendarific.com/api/v2/holidays'
    
    params = {
        'api_key': api_key,
        'country': country_code,
        'year': year
    }
    
    if month:
        params['month'] = month
    if day:
        params['day'] = day
    
    try:
        response = requests.get(url, params=params, timeout=10)
        response.raise_for_status()
        
        data = response.json()
        
        meta = data.get('meta', {})
        if meta.get('code') != 200:
            error_detail = meta.get('error_detail', 'Unknown error')
            code = meta.get('code', 'Unknown')
            raise ValueError(f'Calendarific API error: {code} - {error_detail}')
        
        return normalize_holidays_data(data)
    except requests.RequestException as e:
        raise Exception(f'Failed to fetch holidays data: {str(e)}')
    except (KeyError, ValueError) as e:
        raise Exception(f'Failed to parse holidays data: {str(e)}')


def get_today_holidays(api_key: str, country_code: str) -> HolidaysResponse:
    """
    Gets holidays for today's date
    
    Args:
        api_key: Calendarific API key
        country_code: ISO 3166-1 alpha-2 country code
    
    Returns:
        Holidays data for today
    """
    today = datetime.now()
    year = today.year
    month = today.month
    day = today.day
    
    return get_holidays(api_key, country_code, year, month, day)


def is_holiday(
    api_key: str,
    country_code: str,
    year: int,
    month: int,
    day: int
) -> bool:
    """
    Checks if a specific date is a holiday
    
    Args:
        api_key: Calendarific API key
        country_code: ISO 3166-1 alpha-2 country code
        year: Year
        month: Month (1-12)
        day: Day (1-31)
    
    Returns:
        True if the date is a holiday
    """
    result = get_holidays(api_key, country_code, year, month, day)
    return result['isHoliday']


def normalize_holidays_data(data: dict) -> HolidaysResponse:
    """
    Normalizes Calendarific API response to our standard format
    
    Args:
        data: Raw response from Calendarific API
    
    Returns:
        Normalized holidays response
    """
    response_data = data.get('response', {})
    holidays = response_data.get('holidays', [])
    
    holiday_names = [h.get('name', 'Unknown Holiday') for h in holidays]
    
    normalized_holidays = []
    for holiday in holidays:
        locations_str = holiday.get('locations', '')
        locations = locations_str.split(', ') if locations_str else []
        
        normalized_holidays.append({
            'name': holiday.get('name', 'Unknown Holiday'),
            'description': holiday.get('description', ''),
            'date': holiday.get('date', {}),
            'type': holiday.get('type', ['unknown'])[0] if holiday.get('type') else 'unknown',
            'locations': locations
        })
    
    return {
        'holidays': normalized_holidays,
        'isHoliday': len(holidays) > 0,
        'holidayNames': holiday_names
    }

File: src/services/calendar.js
/**
 * Calendarific API service
 * Provides holiday information to help adjust restaurant recommendations
 * API Documentation: https://calendarific.com/api-documentation
 */

/**
 * @typedef {Object} Holiday
 * @property {string} name - Name of the holiday
 * @property {string} description - Description of the holiday
 * @property {Object} date - Date information
 * @property {string} date.datetime.year - Year
 * @property {string} date.datetime.month - Month
 * @property {string} date.datetime.day - Day
 * @property {string} type - Type of holiday (national, religious, etc.)
 * @property {string[]} locations - Locations where this holiday is observed
 */

/**
 * @typedef {Object} HolidaysResponse
 * @property {Holiday[]} holidays - Array of holidays
 * @property {boolean} isHoliday - Whether the date has any holidays
 * @property {string[]} holidayNames - Names of all holidays on this date
 */

/**
 * Gets holidays for a specific country and date/year
 * @param {string} apiKey - Calendarific API key from environment variable
 * @param {string} countryCode - ISO 3166-1 alpha-2 country code (e.g., 'US', 'GB', 'CA')
 * @param {number} year - Year to get holidays for
 * @param {number} [month] - Optional month (1-12)
 * @param {number} [day] - Optional day (1-31)
 * @returns {Promise<HolidaysResponse>} Holidays data for the specified date/year
 * @throws {Error} If API request fails or API key is missing
 */
async function getHolidays(apiKey, countryCode, year, month = null, day = null) {
  if (!apiKey) {
    throw new Error('Calendarific API key is required. Set CALENDARIFIC_API_KEY environment variable.');
  }

  if (!countryCode || !year) {
    throw new Error('Country code and year are required.');
  }

  const url = new URL('https://calendarific.com/api/v2/holidays');
  url.searchParams.set('api_key', apiKey);
  url.searchParams.set('country', countryCode);
  url.searchParams.set('year', year.toString());
  
  if (month) {
    url.searchParams.set('month', month.toString());
  }
  if (day) {
    url.searchParams.set('day', day.toString());
  }

  try {
    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(
        `Calendarific API error: ${response.status} ${response.statusText}`
      );
    }

    const data = await response.json();

    if (data.meta?.code !== 200) {
      throw new Error(
        `Calendarific API error: ${data.meta?.code || 'Unknown'} - ${data.meta?.error_detail || 'Unknown error'}`
      );
    }

    return normalizeHolidaysData(data);
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Failed to fetch holidays data: ${error.message}`);
    }
    throw new Error('Failed to fetch holidays data: Unknown error');
  }
}

/**
 * Gets holidays for today's date
 * @param {string} apiKey - Calendarific API key
 * @param {string} countryCode - ISO 3166-1 alpha-2 country code
 * @returns {Promise<HolidaysResponse>} Holidays data for today
 */
async function getTodayHolidays(apiKey, countryCode) {
  const today = new Date();
  const year = today.getFullYear();
  const month = today.getMonth() + 1; // getMonth() returns 0-11
  const day = today.getDate();

  return getHolidays(apiKey, countryCode, year, month, day);
}

/**
 * Checks if a specific date is a holiday
 * @param {string} apiKey - Calendarific API key
 * @param {string} countryCode - ISO 3166-1 alpha-2 country code
 * @param {number} year - Year
 * @param {number} month - Month (1-12)
 * @param {number} day - Day (1-31)
 * @returns {Promise<boolean>} True if the date is a holiday
 */
async function isHoliday(apiKey, countryCode, year, month, day) {
  const result = await getHolidays(apiKey, countryCode, year, month, day);
  return result.isHoliday;
}

/**
 * Normalizes Calendarific API response to our standard format
 * @param {Object} data - Raw response from Calendarific API
 * @returns {HolidaysResponse} Normalized holidays response
 */
function normalizeHolidaysData(data) {
  const holidays = data.response?.holidays || [];
  const holidayNames = holidays.map(h => h.name || 'Unknown Holiday');

  return {
    holidays: holidays.map(holiday => ({
      name: holiday.name || 'Unknown Holiday',
      description: holiday.description || '',
      date: holiday.date || {},
      type: holiday.type?.[0] || 'unknown',
      locations: holiday.locations?.split(', ') || []
    })),
    isHoliday: holidays.length > 0,
    holidayNames
  };
}

module.exports = {
  getHolidays,
  getTodayHolidays,
  isHoliday,
  normalizeHolidaysData
};
